From 6f8ef80bb3ba5d244a428aee200c168e809a0079 Mon Sep 17 00:00:00 2001
From: Lennart Poettering <lennart@poettering.net>
Date: Sun, 26 May 2024 00:53:46 +0800
Subject: [PATCH] core/dbus-execute: don't trigger assertion if WorkingDirectory="" or "-"

Follow-up for 14631951cea807de2d482a430841c604c2040718.

Before this commit, if WorkingDirectory= is empty or literally "-",
'simplified' is not populated, resulting in the ASSERT_PTR in
unit_write_settingf() below getting triggered.

Also, do not accept "-", so that the parser is consistent with
load-fragment.c.

Fixes #33015
---
diff --git a/src/core/dbus-execute.c b/src/core/dbus-execute.c
--- a/src/core/dbus-execute.c
+++ b/src/core/dbus-execute.c
@@ -2721,38 +2721,35 @@
 
         } else if (streq(name, "WorkingDirectory")) {
                 _cleanup_free_ char *simplified = NULL;
-                bool missing_ok, is_home;
+                bool missing_ok = false, is_home = false;
                 const char *s;
 
                 r = sd_bus_message_read(message, "s", &s);
                 if (r < 0)
                         return r;
 
-                if (s[0] == '-') {
-                        missing_ok = true;
-                        s++;
-                } else
-                        missing_ok = false;
-
-                if (isempty(s))
-                        is_home = false;
-                else if (streq(s, "~"))
-                        is_home = true;
-                else {
-                        if (!path_is_absolute(s))
-                                return sd_bus_error_set(error, SD_BUS_ERROR_INVALID_ARGS, "WorkingDirectory= expects an absolute path or '~'");
-
-                        r = path_simplify_alloc(s, &simplified);
-                        if (r < 0)
-                                return r;
-
-                        if (!path_is_normalized(simplified))
-                                return sd_bus_error_set(error, SD_BUS_ERROR_INVALID_ARGS, "WorkingDirectory= expects a normalized path or '~'");
-
-                        if (path_below_api_vfs(simplified))
-                                return sd_bus_error_set(error, SD_BUS_ERROR_INVALID_ARGS, "WorkingDirectory= may not be below /proc/, /sys/ or /dev/.");
-
-                        is_home = false;
+                if (!isempty(s)) {
+                        if (s[0] == '-') {
+                                missing_ok = true;
+                                s++;
+                        }
+
+                        if (streq(s, "~"))
+                                is_home = true;
+                        else {
+                                if (!path_is_absolute(s))
+                                        return sd_bus_error_set(error, SD_BUS_ERROR_INVALID_ARGS, "WorkingDirectory= expects an absolute path or '~'");
+
+                                r = path_simplify_alloc(s, &simplified);
+                                if (r < 0)
+                                        return r;
+
+                                if (!path_is_normalized(simplified))
+                                        return sd_bus_error_set(error, SD_BUS_ERROR_INVALID_ARGS, "WorkingDirectory= expects a normalized path or '~'");
+
+                                if (path_below_api_vfs(simplified))
+                                        return sd_bus_error_set(error, SD_BUS_ERROR_INVALID_ARGS, "WorkingDirectory= may not be below /proc/, /sys/ or /dev/.");
+                        }
                 }
 
                 if (!UNIT_WRITE_FLAGS_NOOP(flags)) {
@@ -2760,7 +2757,10 @@
                         c->working_directory_home = is_home;
                         c->working_directory_missing_ok = missing_ok;
 
-                        unit_write_settingf(u, flags|UNIT_ESCAPE_SPECIFIERS, name, "WorkingDirectory=%s%s", missing_ok ? "-" : "", c->working_directory_home ? "+" : ASSERT_PTR(c->working_directory));
+                        unit_write_settingf(u, flags|UNIT_ESCAPE_SPECIFIERS, name,
+                                            "WorkingDirectory=%s%s",
+                                            c->working_directory_missing_ok ? "-" : "",
+                                            c->working_directory_home ? "~" : strempty(c->working_directory));
                 }
 
                 return 1;
